keps/sig-testing/3041-node-conformance-and-features/README.md

```diff
@@ -0,0 +1,803 @@
+  that is not a standard OSS test infrastructure machines needed to run the feature.
+
+### Non-Goals
+
+- As `[NodeConformance]` may be a confusing term and may be confused with the
+  k8s Conformance tests, renaming of it is not in a scope of this KEP.
+- Convert textual tags into the ginkgo v2 tags. This is optional, but not
+  required and can be handled outside of this KEP.
+
+## Proposal
+
+### Clarify the semantic of a `[Feature:]` label
+
+The current [feature definition is](https://github.com/kubernetes/community/blob/32a1c14d04ff78684d78b827ac7c49f70352d509/contributors/devel/sig-testing/e2e-tests.md#kinds-of-tests):
+
+> * `[Feature:.+]`: If a test has non-default requirements to run or targets
+>   some non-core functionality, and thus should not be run as part of the
+>   standard suite, it receives a `[Feature:.+]` label, e.g.
+>   `[Feature:Performance]` or `[Feature:Ingress]`. `[Feature:.+]` tests are not
+>   run in our core suites, instead running in custom suites. If a feature is
+>   experimental or alpha and is not enabled by default due to being incomplete
+>   or potentially subject to breaking changes, it does not block PR merges, and
+>   thus should run in some separate test suites owned by the feature owner(s)
+>   (see Continuous Integration below).
+
+Feature will indicate that things may NOT work on all k8s distros and/or node
+capabilities. This flag will not indicate the Feature Gate state.
+
+Proposed change:
+
+> * `[Feature:.+]`: If a test validates functionality that may only work
+>   outside the minimal conformant installation of Kubernetes, e.g. when
+>   specific node capabilities are enabled, having a certain addons available,
+>   like loadbalancer integration, or when test depends on the
+>   functionality of underlying components like container runtime, and thus may
+>   need to be skipped on certain environments, it receives a `[Feature:.+]`
+>   label, e.g. `[Feature:AppArmor]`. This label can also be applied to signify that
+>   test validates non-core functionality, like `[Feature:Ingress]`.
+>   `[Feature:.+]` tests are not run in a core suites, but typically can run together
+>   on the standard environment as many capabilities and functionality of underlying
+>   components are pre-enabled on the standard environment. Use `[Environment:Foo]` to
+>   signify the specific environment configuration needed to run the test.
+>   For example if test requires GPU to be configured, but tests the non-optional
+>   feature like a device plugin.
+>   `[Feature:+]` label should not be confused with `[FeatureGate:]` label.
+
+And another two:
+
+> * `[FeatureGate:.+]`: If a test only works when the certain feature gate is
+>   enabled it receives a `[FeatureGate:.+]` label. `[FeatureGate:.+]` tests
+>   must also be marked with the status of this feature gate: `[Alpha]`,
+>   `[Beta]`, `[Stable]`, `[Deprecated]`. This label helps to skip tests that
```

Comment:
I see a few problems with requiring these `[Alpha]/[Beta]/[Stable]/[Deprecated]' tags.

The first is conceptual: if a test depends on two different features, what tag should be used? All that apply, e.g. `[FeatureGate:foo] [Beta] [FeatureGate:bar] [Stable]`?

In the past, annotations where removed once a feature graduated to stable, which then also automatically enabled the test in the core jobs. Is the plan now to keep the `[FeatureGate]` annotation and instead adapt job definitions when a feature graduates to GA? I can see that it would be valuable to keep the `[FeatureGate]` annotation because it provides information that otherwise would get lost. For example, one might want to run all tests for a specific feature even after it has graduated. I am less sure that `[Stable]` is useful.

My biggest concern is that the `[Alpha]/[Beta]/[Stable]/[Deprecated]` annotations are redundant and that maintaining them is going to be error-prone.

Instead of adding them manually to the test description, can we do that with helper functions? Here's how that could look like:
```
Current solution:
ginkgo.It("foo bar [FeatureGate:foo] [Beta]", ...)

Revised approach:
framework.It("foo bar", framework.WithFeatureGate("foo"), ...)
```

`framework.It` then acts as wrapper around `ginkgo.It`. When one of the optional parameters is the result of `framework.WithFeatureGate`, it does several things:
- Checks that the feature is registered in a feature gate. That feature gate could be seeded with the Kubernetes features, but additional features could be added.
- Retrieves the stability level of that feature from the feature gate.
- Extends the test description with, in this example, `[FeatureGate:foo] [Beta]`.
- Add `ginkgo.Label("FeatureGate:foo")` and `ginkgo.Label("Beta")` when calling `ginkgo.It`.

Another advantage is that typos in the feature name will be detected when registering the tests, i.e. when the suite starts. When embedded in the text we have to rely on code reviews.

Similar helper functions for `[Environment]` and `[Feature]` could also be added. Then it becomes possible to select tests with:
```
  --label-filter [expression] 
    If set, ginkgo will only run specs with labels that match the label-filter. 
    The passed-in expression can include boolean operations (!, &&, ||, ','),
    groupings via '()', and regular expressions '/regexp/'.  e.g. '(cat || dog)
    && !fruit'
```

Reply:
> The first is conceptual: if a test depends on two different features, what tag should be used? All that apply, e.g. `[FeatureGate:foo] [Beta] [FeatureGate:bar] [Stable]`?

we should not have features that depends on features that are not stable, right?

> I can see that it would be valuable to keep the [FeatureGate] annotation because it provides information that otherwise would get lost

I see that more as part of the `[Feature:]` tag, feature-gates should be part of the system when they graduate

Reply:
By requiring that all features, feature gates and environments must be registered in the framework, we can add command line flags that list them. Might be useful.

Reply:
yeah, agree, we can do much better and we should aim for that, but at this point I think we should fix current state and have the right variables, and then iterate on those.
It seems this KEP allows us to define those variables with Feature, FeatureGate and Environment ... if we agree on those, we can start working later no how to "enforce" those and how to improve the automation , WDYT?

Reply:
See https://github.com/kubernetes/kubernetes/pull/112894/files for a prototype implementation of this approach.

Reply:
> we should not have features that depends on features that are not stable, right?

Yes, but a test itself might depend on two independent features. It's unlikely, I just wanted to call it out.

> I see that more as part of the [Feature:] tag, feature-gates should be part of the system when they graduate, if the FeatureGate enables a new Feature we can use this tag

So before a feature is GA, tests get annotated with `[FeatureGate:foo] [Beta]`, then when they graduate, we replace that with `[Feature:foo] [Stable]`? That seems a bit inconsistent to me.

> at this point I think we should fix current state and have the right variables, and then iterate on those

I'm worried about code churn and reviewer fatigue.If we do this in two steps, we have to touch all test definitions twice, first to clean up the text, then later to turn the text into function calls. Reviewing the first step will be harder because typos have to be caught by reviewers.

Reply:
> So before a feature is GA, tests get annotated with [FeatureGate:foo] [Beta], then when they graduate, we replace that with [Feature:foo] [Stable]? That seems a bit inconsistent to me.

I think that this is one of the problems we are having, FeatureGates and Feature are different things, the problem that I see is that sometimes FeatureGates are used to gate Features, but not all FeatureGates are Features

> I'm worried about code churn and reviewer fatigue.If we do this in two steps, we have to touch all test definitions twice, first to clean up the text, then later to turn the text into function calls. Reviewing the first step will be harder because typos have to be caught by reviewers.

I think that is a fantastic idea to use the functions as you described and keep it in sync with the code ... but I don't want to impose it to the KEP author 😄

Reply:
I can help with using those functions, both by updating the KEP and providing the implementation.

But let's clarify first how to handle graduated features. That has implications that are unclear, both with and without those helper functions.